Initial APIs for storing E2E room keys

ara4n opened the pull request:
Implementing the proposal in https://docs.google.com/document/d/1MOoIA9qEKIhUQ3UmKZG-loqA8e0BzgWKKlKRUGMynVc/edit#

Now ready for final review

erikjohnston commented on synapse/handlers/e2e_room_keys.py:
```diff
@@ -0,0 +1,134 @@
+
+                    yield self._upload_room_key(
+                        user_id, version, room_id, session_id, room_key
+                    )
```
I'd be tempted to do some of the different rooms concurrently, given its in a lock.

ara4n replied:
i think this is preemie optimisation; we don't even have a flow yet where the bulk upload is used. so have added a comment instead.

erikjohnston commented on synapse/handlers/e2e_room_keys.py:
```diff
@@ -0,0 +1,134 @@
+                raise e
+
+        if version_info['version'] != version:
+            raise RoomKeysVersionError(current_version=version_info.version)
```
Don't we need to lock to ensure that we don't change version while doing the checks?

ara4n replied:
mmm. fixed.

erikjohnston commented on synapse/handlers/e2e_room_keys.py:
```diff
@@ -0,0 +1,134 @@
+class E2eRoomKeysHandler(object):
+    def __init__(self, hs):
+        self.store = hs.get_datastore()
+        self._upload_linearizer = Linearizer("upload_room_keys_lock")
```
Might be worth commenting on what this protects

ara4n replied:
done

erikjohnston commented on synapse/handlers/e2e_room_keys.py:
```diff
@@ -0,0 +1,134 @@
+            elif room_key['forwarded_count'] < room_key['forwarded_count']:
+                pass
+            else:
+                upsert = False
```
I'd be tempted to split this logic out into a freestanding function, as its business logic that could be easily tested. It also gives a chance to document whats going on a bit.

ara4n replied:
done.

erikjohnston commented on synapse/rest/client/v2_alpha/room_keys.py:
```diff
@@ -0,0 +1,279 @@
+         * older first_message_index always win over newer sessions
+         * lower forwarded_count always wins over higher forwarded_count
+
+        We trust the clients not to lie and corrupt their own backups.
```
Presumably this means that if I get your access token I can delete your history?

ara4n replied:
it means you can delete the backup, yup.

erikjohnston commented on synapse/rest/client/v2_alpha/room_keys.py:
```diff
@@ -0,0 +1,279 @@
+
+        We trust the clients not to lie and corrupt their own backups.
+
+        POST /room_keys/keys/!abc:matrix.org/c0ff33?version=1 HTTP/1.1
```
I'm wondering whether we really want the version to be a query param. Whats the rationale there?

ara4n replied:
I briefly considered having /room_keys/keys/$version/$room_id/$session_id as the URI, but decided not to because it implied to me that the schema could change between different versions - whereas in practice a given /room/_keys/keys/$room_id/$session_id URI should have the same content irrespective of the version, which just describes the encryption key. This rationale is pretty ropey though.

@erikjohnston Think it's worth changing to embed it in the path instead?

erikjohnston commented on synapse/rest/client/v2_alpha/room_keys.py:
```diff
@@ -0,0 +1,279 @@
+
+        info = yield self.e2e_room_keys_handler.get_version_info(
+            user_id, version
+        )
```
Should we be checking that version was actually supplied? Would it make more sense to spit this class in two, one doing `POST /room_keys/version` and the other `GET/DELETE /room_keys/version/<version>`?

In fact, having POST <verb> might makes things a bit clearer, e.g. `POST /room_keys/create_version`

ara4n replied:
I didn't bother checking that version was supplied; i think it will explode fairly sensibly if it isn't?

Not sure what the point of splitting into POST and GET/DELETE would be?  And shouldn't POST have the same path anyway? (although ugh rest https://medium.com/@pakaldebonchamp/rest-is-the-new-soap-97ff6c09896d)

erikjohnston replied:
> I didn't bother checking that version was supplied; i think it will explode fairly sensibly if it isn't?

My hunch is that it'd 500 in a fairly non obvious way tbh, or possibly succeed as the database ends up executing `SELECT * FROM versions WHERE version = null`. Not hugely friendly.

> Not sure what the point of splitting into POST and GET/DELETE would be?

So you have a GET and DELETE that only work if a version is supplied, and a POST that only works if you *don't* supply a version. To save having to do checks everywhere, and to make it clearer when a version is required or disallowed, having this as separate classes seems much clearer.

> And shouldn't POST have the same path anyway? 

They don't really share a path as it is. If you think that `POST /room_keys/version` is clearer than `POST /room_keys/create_version` then fair enough, twas but a thought.

And yes, turns out naming APIs is hard.

ara4n replied:
I've changed the semantics so that if version is missing it gives you the most recent version (and 404 if there are none).

I've experimented with building /version into the path, but it feels pretty clunky (given you'd need to know the version in order to know the path, meaning that things like 'give me the current version' are harder to express).

I've also thought about whether `POST /room_keys/version` returning an ID which can then be used with `GET/DELETE /room_keys/version/$id` is bad form, but from my pov this is a pretty standard RESTy pattern.  Especially as we now also have `GET /room_keys/version` to return the current version.  Also, I think this is increasingly dubiously subjective, and in practice as long as it works I'm not sure the precise API shape is that critical unless it's spectacularly inconsistent with the wider CS API.

erikjohnston replied:
Honestly, the thought of changing the API shape was simply a tangential suggestion. The actual suggestion was to change the code layout to better avoid having to manually handle null values of `version`.

Having `GET /version` that returns the most recent version works, but what does it return if there isn't a version for that user? What does `DELETE /version` do? I note that we still don't seem to be checking if the version is null in that case.

ara4n replied:
for `GET /version` when there is no most recent version, you should get a 404, as the docstring says.

I've fixed `DELETE /version` to delete the current backup version info, for symmetry with the rest of the API, and added a test for it.

erikjohnston commented on synapse/storage/e2e_room_keys.py:
```diff
@@ -0,0 +1,240 @@
+                lock=False,
+            )
+
+            return True
```
Isn't there a `_simple_upsert`?

ara4n replied:
so there is.

erikjohnston commented on synapse/storage/e2e_room_keys.py:
```diff
@@ -0,0 +1,240 @@
+
+    # XXX: this isn't currently used and isn't tested anywhere
+    # it could be used in future for bulk-uploading new versions of room_keys
+    # for a user or something though.
```
Unless you're planning on implementing this I'd rather we just got rid of it, especially if it isn't tested.

ara4n replied:
i'd rather not lose the code on the floor, given it may well be useful in future.

erikjohnston replied:
I'd rather not have to review code that isn't used, and dead code just adds bloat to the code base and makes understanding things more annoying.

Also, I think I would ask you to rewrite.

ara4n replied:
ok, done

erikjohnston commented on synapse/storage/e2e_room_keys.py:
```diff
@@ -0,0 +1,240 @@
+                    return dict.__getitem__(self, item)
+                except KeyError:
+                    value = self[item] = type(self)()
+                    return value
```
Not sure this will work if you have a mix of `AutoVivification` and `dict`, tbh.

erikjohnston replied:
Oh, no, I misread how you were using it and it probably does work.

erikjohnston commented on synapse/storage/e2e_room_keys.py:
```diff
@@ -0,0 +1,240 @@
+
+        sessions = AutoVivification()
+        for row in rows:
+            sessions['rooms'][row['room_id']]['sessions'][row['session_id']] = {
```
I think this can be written as:

```python
sessions['rooms'].setdefault(row['room_id'], {"sessions":{}})['sessions'][row['session_id']] = { ... }
```

Or if  you don't mind that we use an extra line:

```python
room_entry = sessions['rooms'].setdefault(row['room_id'], {"sessions":{}})
room_entry['sessions'][row['session_id']] = { ... }
```

ara4n replied:
thanks; done

erikjohnston commented on synapse/storage/e2e_room_keys.py:
```diff
@@ -0,0 +1,240 @@
+        if room_id:
+            keyvalues['room_id'] = room_id
+        if session_id:
+            keyvalues['session_id'] = session_id
```
If `room_id` and `session_id` aren't required I think it would be polite to make them optional arguments, even if that isn't used. It makes it a lot clearer just from the signature line that the params are optional.

erikjohnston replied:
Also, does it make sense to have a session_id without a room_id?

ara4n replied:
fixed

erikjohnston commented on synapse/storage/schema/delta/46/e2e_room_keys.sql:
```diff
@@ -0,0 +1,40 @@
+
+CREATE UNIQUE INDEX e2e_room_keys_user_idx ON e2e_room_keys(user_id);
+CREATE UNIQUE INDEX e2e_room_keys_room_idx ON e2e_room_keys(room_id);
+CREATE UNIQUE INDEX e2e_room_keys_session_idx ON e2e_room_keys(session_id);
```
Ermh, are you sure that you want these to be unique indices? What if you have multiple users uploading keys for the same room?

erikjohnston replied:
Looking at the code I think we can get away with a single index of `(user_id, room_id, session_id)`, which will be much faster.

ara4n replied:
oops, fixed.

erikjohnston commented on synapse/storage/schema/delta/46/e2e_room_keys.sql:
```diff
@@ -0,0 +1,40 @@
+    auth_data TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX e2e_room_key_user_idx ON e2e_room_keys(user_id);
```
This is not what you mean, I assume:

```sql
CREATE UNIQUE INDEX e2e_room_key_versions_idx ON e2e_room_key_versions(user_id);
```

ara4n replied:
oops, fixed

erikjohnston commented on synapse/storage/schema/delta/46/e2e_room_keys.sql:
```diff
@@ -0,0 +1,40 @@
+CREATE UNIQUE INDEX e2e_room_keys_session_idx ON e2e_room_keys(session_id);
+
+-- the metadata for each generation of encrypted e2e session backups
+CREATE TABLE e2e_room_key_versions (
```
Room keys? There isn't a room_id in the table.

ara4n replied:
fixed to be called `e2e_room_keys_versions` which is more grammatically correct, given we are talking about versions of sets of backup of multiple room_keys here.

ara4n commented on synapse/api/errors.py:
```diff
@@ -233,6 +234,22 @@ def error_dict(self):
+    """A client has tried to upload to a non-current version of the room_keys store
+    """
+    def __init__(self, code=403, msg="Wrong room_keys version", current_version=None,
+                 errcode=Codes.WRONG_ROOM_KEYS_VERSION):
```
have made `code`, `msg` & `errcode` unconfigurable, and de-optionalised current_version.

ara4n commented on synapse/handlers/e2e_room_keys.py:
```diff
@@ -0,0 +1,134 @@
+
+
+class E2eRoomKeysHandler(object):
+    def __init__(self, hs):
```
done

ara4n commented on synapse/handlers/e2e_room_keys.py:
```diff
@@ -0,0 +1,134 @@
+
+    @defer.inlineCallbacks
+    def delete_room_keys(self, user_id, version, room_id, session_id):
+        yield self.store.delete_e2e_room_keys(user_id, version, room_id, session_id)
```
oops; added for consistency with uploads now

ara4n commented on synapse/handlers/e2e_room_keys.py:
```diff
@@ -0,0 +1,134 @@
+            # go through the room_keys
+            for room_id in room_keys['rooms']:
+                for session_id in room_keys['rooms'][room_id]['sessions']:
+                    room_key = room_keys['rooms'][room_id]['sessions'][session_id]
```
done

erikjohnston commented on synapse/storage/e2e_room_keys.py:
```diff
@@ -0,0 +1,227 @@
+    def get_e2e_room_keys(
+        self, user_id, version, room_id=None, session_id=None
+    ):
+
```
Its probably worth documenting what happens in the presence of a null room_id and null session_id

ara4n replied:
done

erikjohnston commented on synapse/storage/e2e_room_keys.py:
```diff
@@ -0,0 +1,227 @@
+    def delete_e2e_room_keys(
+        self, user_id, version, room_id=None, session_id=None
+    ):
+
```
Ditto with docstrings

ara4n replied:
done

ara4n commented:
@erikjohnston I've:
 * added full docstring
 * added UTs
 * addressed all your PR feedback (i think), to date
If you can give a final check it'd be much appreciated; merry new year...

In other news (cc @richvdh @erikjohnston), this felt *really* painful to write - and not just because I'm a bit rusty.

In the end, this is an incredibly trivial API, for simple CRUD of some JSON objects.  The only subtlety is the merge semantics when updating, but that's a whole 5 lines of code.  But meanwhile, the overall PR has ended up being a 1,511 line diff (including tests), which feels insanely heavyweight for a relatively non-verbose language like Python.  Stuff which felt particularly "gah we're doing this wrong" included:
 * The triple-layering of rest, handler & storage.  It feels like the rest layer is spectacularly premature optimisation, given rest is the only transport synapse supports, and it doubles all the docstring and threading the API signatures through the stack - lots of overhead and boilerplate for pretty much zero value.
 * The duplication of docstring doc between all three layers.
 * The fact we don't have any generic ways for storing JSON in DB tables (a la glow) which this could have utilised in the first place.
 * The fact that docstring is completely freeform, and nothing provides any help in ensuring its correctness (let alone the pep-484 type hinting)
 * The fact our linting seems particularly insanely stringent - although this may be my fault for not using a pep-8 aware editor.  It's very frustrating for the build to fail due to having trailing whitespace on a blank line.

Anyway, though I'd mention this, as if it's painful for me, I imagine it's also painful for other contributors...

erikjohnston commented on synapse/storage/e2e_room_keys.py:
```diff
@@ -0,0 +1,288 @@
+                    "SELECT MAX(version) FROM e2e_room_keys_versions WHERE user_id=?",
+                    (user_id,)
+                )
+                this_version = txn.fetchone()[0]
```
What if there isn't a version for this userID

ara4n replied:
fixed

erikjohnston commented on synapse/storage/e2e_room_keys.py:
```diff
@@ -0,0 +1,288 @@
+        keyvalues = {
+            "user_id": user_id,
+            "version": version,
+        }
```
Why is this a separate variable?

ara4n replied:
no idea. fixed.

erikjohnston commented on synapse/rest/client/v2_alpha/room_keys.py:
```diff
@@ -0,0 +1,351 @@
+
+        requester = yield self.auth.get_user_by_req(request, allow_guest=False)
+        user_id = requester.user.to_string()
+        version = request.args.get("version")[0]
```
For query params its probably better to use [`parse_string`](https://github.com/matrix-org/synapse/blob/552f123bea1014680ab798b7e34cd1b23424a189/synapse/http/servlet.py#L105) and co

ara4n replied:
done

erikjohnston commented:
TBH I wouldn't classify this as simple CRUD, given the merging semantics and linearization that is going on. 

> The triple-layering of rest, handler & storage. It feels like the rest layer is spectacularly premature optimisation, given rest is the only transport synapse supports, and it doubles all the docstring and threading the API signatures through the stack - lots of overhead and boilerplate for pretty much zero value.

The separation has been really quite useful to split out the parsing of the request versus the actual functionality. Merging rest/handlers quickly become unwieldy in most cases, and it makes skimming the rest layer to figure out what the APIs are really easy, which is something I do a lot to get back up to speed with how some bit of synapse works.

> The duplication of docstring doc between all three layers.

This is a nuisance, yes, though its perhaps worth noting that the documentation on the rest layer is describing the APIs in detail, while the handler functions are describing exactly what each bit is doing under the hood.

In general I've found that even without explicit layering you end up with duplicated documentation, unless you have a single mammoth function that does everything.

> The fact we don't have any generic ways for storing JSON in DB tables (a la glow) which this could have utilised in the first place.

The structure of the glow database was quite complicated, iirc, as you need to somehow magically have all the correct indices in place. It also didn't support common queries, e.g. `MAX(..)` which you are using.

If you simply want to store JSON in the database then that's easy─and we do it all over the place─just use `json.dumps(..)` and `json.loads(..)` and have a `json` column in the table.

Other than not requiring you to define a table and what indices you want, I'm not sure that it would actually buy you very much here, since you'd still have all the merging and validation logic.

> The fact our linting seems particularly insanely stringent - although this may be my fault for not using a pep-8 aware editor. It's very frustrating for the build to fail due to having trailing whitespace on a blank line.

The flip side is that it makes the code so much nicer to work on. Most developers nowadays are so used to having linters that its second nature.

richvdh commented:
> In other news (cc @richvdh @erikjohnston), this felt really painful to write - and not just because I'm a bit rusty.

Not much to add to your pain points beyond what erik has already written here, all of which I agree with completely.

FWIW I see a 1366 line diff rather than 1511, of which a quarter is tests (for which yay), and, eyeballing it, it looks like fully half of it is either doc-comments (for which yay again) or copyright notices (meh).

I think that one of the reasons that synapse has been so hard to work on in the past is that we have not done well with doc-comments, which leads to a lot of spelunking if you want to change existing code. The fact that this code is now clearly documented is something to be proud of, not be sad about.

ara4n commented:
@erikjohnston ptal.

ara4n commented:
Ironically, there are some problems which have turned up with this whilst implementing the clientside, so I'm going to avoid merging this until I'm sure it's on the right track.  Need to:

 * [ ] Need to include ephemeral public key alongside session_data in order to calculate the AES key
 * [ ] Need to Include backup public key alongside the version info, so it can be passed to new devices

richvdh commented:
let's reopen this when it's ready, then

uhoreg commented on synapse/storage/e2e_room_keys.py:
```diff
@@ -0,0 +1,308 @@
+class EndToEndRoomKeyStore(SQLBaseStore):
+
+    @defer.inlineCallbacks
+    def get_e2e_room_key(self, user_id, version, room_id, session_id):
```
This function is only used when checking to see if we should overwrite a key.  Would it be better to just convert that bit of code to use `get_e2e_room_keys` instead, and get rid of this function?
